ruler/scheduling_queue.go

```diff
@@ -0,0 +1,187 @@
+package ruler
+
+import (
+	"container/heap"
+	"time"
+
+	"github.com/jonboulle/clockwork"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+var (
+	queueLength = prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: "cortex",
+		Name:      "rules_queue_length",
+		Help:      "The length of the rules queue.",
+	})
+)
+
+func init() {
+	prometheus.MustRegister(queueLength)
+}
+
+// ScheduledItem is an item in a queue of scheduled items.
+type ScheduledItem interface {
+	Key() string
+	// Scheduled returns the earliest possible time the time is available for
+	// dequeueing.
+	Scheduled() time.Time
+}
+
+type item struct {
+	index   *int
+	payload ScheduledItem
+}
+
+type scheduledItems []*item
+
+func (q scheduledItems) Len() int { return len(q) }
```

Comment:
As discussed, make the thing satisfying `heap.Interface` the `queueState` not this type - in fact I think this type could go away.